File: c/INA260.c
#include "INA260.h"
#include "smbus.h"
#include <fcntl.h>
#include <unistd.h>
#include <linux/i2c-dev.h>
#include <linux/i2c.h>
#include <sys/ioctl.h>
#include <stdio.h>
#include <math.h>
#define VERBOSE 0

__u8 rd_err = 0;
__u8 wr_err = 0;


int i2c_init(__u8 dev_addr)
{
	// Returns a file id
	int fd = 0;
	char *fileName = "/dev/i2c-1";
	
	// Open port for reading and writing
	if ((fd = open(fileName, O_RDWR)) < 0)
	{
		if (VERBOSE) printf("Error! Cannot opend the port\n");
	}
	
	// Set the port options and set the address of the device
	if (ioctl(fd, I2C_SLAVE, dev_addr) < 0) 
	{					
		close(fd);
		if (VERBOSE) printf("Device with device address  %02x is not reachable\n",dev_addr);
	}

	return fd;
}

__u16 bitset(__u16 number, __u8 i)
{
	 /*
	Sets i-th bit of "number" to 1. i
	Parameters
		number: the input data
		i: contains the location of the bit that needs to be set

	Returns the updated data
	*/
	__u16 output = number;
	output = output | (0x0001 << i);
	return output;
}

__u16 bitclear(__u16 number, __u8 i)
{
	/*
	Sets i-th bit of "number" to 0. i 
	Parameters
		number: the input data
		i: contains the location of bit that needs to be cleared

	Returns the updated data
	 */	
	__u16 output = number;
	output = output & (~(0x0001 << i));
	return output;
	
}


__s8 ina260_config(int fd, __u8 current_enable, __u8 voltage_enable, int converstion_time)
{
	/*
	Configures the INA260 registers

	Returns 0 if the configuration is succesfull.
	*/
	// Reset the device
	__s8 status = 0;
	__u8 wr_addr = REG_CONFIG;
	__u16 wr_data = 0x0000;
	wr_data = bitset(wr_data, RST);
	write_reg(fd, wr_addr, wr_data, &wr_err);

	usleep(20000);

	// Check if device register reading is working
	__u16 man_id = manufacturer_id(fd);
	if (man_id != MAN_ID)
	{
		if (VERBOSE) printf("Device is not reachable. Wrong manufacturer ID\n");
		status = status | (1<<1);
	}

	__u16 die_id_reg = die_id(fd);
	if (die_id_reg != DIE_ID)
	{
		if (VERBOSE) printf("Device is not reachable. Wrong Die ID\n");
		status = status | (1<<2);
	}

	// Set Confugation Register
	wr_addr = REG_CONFIG;
	wr_data = 0x0000;
	
	// Check Datasheet of INA260 for each of the bits
	
	// Setting the read-only bits! (CONF_ROX)
	wr_data = bitset(wr_data, CONF_RO2);
	wr_data = bitset(wr_data, CONF_RO1);
	
	// 
	// Enabling current and voltage based on user input
	if (current_enable==1)
		wr_data = bitset(wr_data, MODE0);

	if (voltage_enable==1)
		wr_data = bitset(wr_data, MODE1);
	
	// Enabling continuous conversion mode
	wr_data = bitset(wr_data, MODE2);

	// Setting conversion time
	switch (converstion_time)
	{
		case 140:
			break;
		case 204:
			bitset(wr_data, ISHCT0);
			bitset(wr_data, VBUSCT0);
			break;
		case 332:
			bitset(wr_data, ISHCT1);
			bitset(wr_data, VBUSCT1);
			break;
		case 588:
			bitset(wr_data, ISHCT0);
			bitset(wr_data, VBUSCT0);
			bitset(wr_data, ISHCT1);
			bitset(wr_data, VBUSCT1);
			break;
		case 1100:
			bitset(wr_data, ISHCT2);
			bitset(wr_data, VBUSCT2);
			break;
		case 2116:
			bitset(wr_data, ISHCT0);
			bitset(wr_data, VBUSCT0);
			bitset(wr_data, ISHCT2);
			bitset(wr_data, VBUSCT2);
			break;
		case 4156:
			bitset(wr_data, ISHCT1);
			bitset(wr_data, VBUSCT1);
			bitset(wr_data, ISHCT2);
			bitset(wr_data, VBUSCT2);
			break;
		case 8244:
			bitset(wr_data, ISHCT0);
			bitset(wr_data, VBUSCT0);
			bitset(wr_data, ISHCT1);
			bitset(wr_data, VBUSCT1);
			bitset(wr_data, ISHCT2);
			bitset(wr_data, VBUSCT2);
			break;
		default:
			break;
	}	

	// Performing register write operation
	write_reg(fd, wr_addr, wr_data, &wr_err);

	usleep(20000);

	// Checking if the register is properly set
	__u16 rd_data = read_reg(fd, wr_addr, &rd_err);
	if (rd_data != wr_data)
	{
		if (VERBOSE) printf("Device configuration failed\n");
		status = status | (1<<3);
	}
	
	return status;

}

__u16 read_reg(int fd, __u8 address, __u8* err)
{
	/*
		Reads a word from the device
	
	Parameters:
		address: register address

	Returns the register value (16 bits)
	*/
	__s32 res = i2c_smbus_read_word_data(fd, address);
	*err = 0;
	if (res < 0) 
	{
		if (VERBOSE) printf("Error in reading! Error code %08X \n",res);
		*err = 1;
		close(fd);
	}

	// Convert result to 16 bits and swap bytes
	res = ((res<<8) & 0xFF00) | ((res>>8) & 0xFF);

	return res;
}

void write_reg(int fd, __u8 address, __u16 data, __u8* err)
{
	/*
        Writes a word to the device
        
        Parameters:
            address: register address
            data: resgister data (16 bits)
	*/	
	*err = 0;
	__s32 res = i2c_smbus_write_word_data(fd, address, ((data<<8) & 0xFF00) | ((data>>8) & 0xFF));
	if (res < 0) 
	{
		if (VERBOSE) printf("Error in writing! Error code %08X \n",res);
		*err = 1;
		close(fd);
	}
	if (VERBOSE) printf("Written address: %02X \t   Written data: %04X\n",address,data);
}

__u16 voltage_read(int fd)
{
	// Returns the voltage register of INA260 (Register 0x02)
	__u16 output = read_reg(fd, REG_BUS_VOLTAGE, &rd_err);
	return rd_err?0x7FFF:output;
}

__s16 reg_to_volt(__u16 reg_voltage_raw)
{
	/*
	Converts the voltage register raw value to Millivolts
	Parameters:
		reg_voltage_raw: raw value read from voltage register of INA260

	Returns the voltage in millivolts
	*/
	return round(((__s16)reg_voltage_raw)*1.25);  //   1.25mv/bit
}

__u16 current_read(int fd)
{
	// Returns the current register of INA260 (Register 0x01)
	__u16 output = read_reg(fd, REG_CURRENT, &rd_err);
	return rd_err?0x7FFF:output;
}

__s16 reg_to_amp(__u16 reg_current_raw)
{
	/*
	Converts the current register raw value to Miillimpers
	
	Parameters:
		reg_current_raw: raw value read from current register of INA260

	Returns the Current in milliampers
	*/
	__u32 current_raw_32 = reg_current_raw;
	__s16 current;
	if (current_raw_32 & (1 << 15)) //Two's complement
		current = (current_raw_32 - 65535);
	else
		current = current_raw_32;

	return round((current*1.25));  //   1.25mA/bit
}

__u16 power_read(int fd)
{
	// Returns the power register of INA260 (Register 0x03)
	__u16 output = read_reg(fd, REG_POWER, &rd_err);
	return rd_err?0x7FFF:output;
}

__u16 reg_to_watt(__u16 reg_power_raw)
{
	/*
	Converts the power register raw value to Milliwatts
	Parameters:
		reg_power_raw: raw value read from power register of INA260

	Returns the Power in milliwatts
	*/
	return reg_power_raw*10;  //   10mW/bit
}


__u16 manufacturer_id(int fd)
{
	/*
    Returns the manufacturer ID - it should always be 0x5449
	*/
	return read_reg(fd, REG_MANUFACTURER_ID, &rd_err);
}

__u16 die_id(int fd)
{
	/*
        Returns the die ID register - it should be 0x2270.
	*/
	return read_reg(fd, REG_DIE_ID, &rd_err);
}

File: python/ina260.py
from smbus2 import SMBus
import time

PCA_AUTOINCREMENT_OFF = 0x00
PCA_AUTOINCREMENT_ALL = 0x80
PCA_AUTOINCREMENT_INDIVIDUAL = 0xA0
PCA_AUTOINCREMENT_CONTROL = 0xC0
PCA_AUTOINCREMENT_CONTROL_GLOBAL = 0xE0

REG_CONFIG = 0x00
REG_CURRENT = 0x01
REG_BUS_VOLTAGE = 0x02
REG_POWER = 0x03
REG_MASK_ENABLE = 0x06
REG_ALERT = 0x07
REG_MANUFACTURER_ID = 0xFE
REG_DIE_ID = 0xFF

MAN_ID = 0x5449
DIE_ID = 0x2270

RST = 15
CONF_RO2 = 14
CONF_RO1 = 13
CONF_RO0 = 12
AVG2 = 11
AVG1 = 10
AVG0 = 9
VBUSCT2 = 8
VBUSCT1 = 7
VBUSCT0 = 6
ISHCT2 = 5
ISHCT1 = 4
ISHCT0 = 3
MODE2 = 2
MODE1 = 1
MODE0 = 0

OCL = 15
UCL = 14
BOL = 13
BUL = 12
POL = 11
CNVR = 10
AFF = 4
CVRF = 3
OVF = 2
APOL = 1
LEN = 0

class INA260:
       
    def __init__(self, address=0x40, channel=1):
        self.i2c_channel = channel
        self.bus = SMBus(self.i2c_channel)
        self.address = address

    def byte(self, number, i):
        """
        Returns the i-th byte of "number"
        Parameters
            number: the input data
            i: contains the byte location to be retrieved

        Returns the i-th byte of "number"
        """
        return (number & (0xff << (i * 8))) >> (i * 8)

    def bitset(self, number, i):
        """
        Sets i-th bit of "number" to 1. i can be a list 
        Parameters
            number: the input data
            i: contains the location of all the bits that need to be set

        Returns the updated data
        """
        output_number = number
        if isinstance(i, list):
            for num in i:
                output_number = output_number | (1 << num)
        else:
            output_number = output_number | (1 << i)
        return output_number
        
    def bitclear(self, number, i):
        """
        Sets i-th bit of "number" to 0. i can be a list 
        Parameters
            number: the input data
            i: contains the location of all the bits that need to be cleared

        Returns the updated data
        """
        output_number = number
        if isinstance(i, list):
            for num in i:
                output_number = output_number & ~(1 << num)
        else:
            output_number = output_number & ~(1 << i)
        return output_number
    def config(self):
        """
        Configures the INA260 registers

        Returns 0 if the configuration is succesfull.
        """
        status = 0
        # Reset the device
        wr_addr = REG_CONFIG
        wr_data = 0x0000
        wr_data = self.bitset(wr_data, RST)
        try:
            self.write_reg(wr_addr, wr_data)
        except:
            status |= (1<<0)
            print("Unable to connect to the device")
            return status
        
        time.sleep(0.1)

        # Check if device register reading is working 
        man_id = self.manufacturer_id()
        if man_id != MAN_ID:
            print("Device is not reachable. Wrong manufacturer ID")
            status |= (1<<1)
            return status
        die_id = self.die_id()
        if die_id != DIE_ID:
            print("Device is not reachable. Wrong Die ID")
            status |= (1<<2)
            return status
        
        # Set Confugation Register
        wr_addr = REG_CONFIG
        wr_data = 0x0000
        # Check Datasheet of INA260 for each of the bits
        # Make sure to set the read-only bits!!! (CONF_ROX)
        wr_data = self.bitset(wr_data, [CONF_RO2, CONF_RO1, MODE2, MODE0]) 
        # wr_data = self.bitclear(wr_data, [CONF_RO0, AVG2, AVG1, AVG0, VBUSCT2, VBUSCT1, VBUSCT0, ISHCT2, ISHCT1, ISHCT0, MODE1])
        self.write_reg(wr_addr, wr_data)
        time.sleep(0.1)
        rd_data = self.read_reg(wr_addr)
        if rd_data != wr_data:
            print("Device configuration failed")
            status |= (1<<3)
        return status



    def read_reg(self, reg):
        """
        Reads a word from the device
        
        Parameters:
            reg: register address

        Returns the register value (16 bits)
        """
        reg_val = self.bus.read_i2c_block_data(self.address, reg, 2)
        return reg_val[1] | (reg_val[0]<<8)

    def write_reg(self, reg, data):
        """
        Writes a word to the device
        
        Parameters:
            reg: register address
            data: resgister data (16 bits)
        """
        write_data = [self.byte(data,1), self.byte(data,0)]
        res = self.bus.write_i2c_block_data(self.address, reg, write_data)

    def voltage_reg(self):
        """
        Returns the voltage register of INA260 (Register 0x02)
        """
        reg_voltage_raw = self.read_reg(REG_BUS_VOLTAGE)
        return reg_voltage_raw

    def reg_to_volt(self, reg_voltage_raw):
        """
        Converts the voltage register raw value to Volt
        Parameters:
            reg_voltage_raw: raw value read from voltage register of INA260

        Returns the voltage in Volt
        """
        voltage = reg_voltage_raw*0.00125 # 1.25mv/bit
        return voltage

    def current_reg(self):
        """
        Returns the current register of INA260 (Register 0x01)
        """
        reg_current_raw = self.read_reg(REG_CURRENT)
        return reg_current_raw

    def reg_to_amp(self, reg_current_raw):
        """
        Converts the current register raw value to Ampers
        
        Parameters:
            reg_current_raw: raw value read from current register of INA260

        Returns the Current in Ampers
        """
        if reg_current_raw & (1 << 15): #Two's complement
            current = reg_current_raw - 65535
        else:
            current = reg_current_raw
            
        current *=0.00125 # 1.25mA/bit
        return current
    def power_reg(self):
        reg_power_raw = self.read_reg(REG_POWER)
        return reg_power_raw

    def reg_to_watt(self, reg_power_raw):
        """
        Converts the power register raw value to Watts
        Parameters:
            reg_power_raw: raw value read from power register of INA260

        Returns the Power in Watts
        """
        power = reg_power_raw* 0.01 # 10mW/bit
        return power

    def manufacturer_id(self):
        """
        Returns the manufacturer ID - it should always be 0x5449
        """
        man_id = self.read_reg(REG_MANUFACTURER_ID)
        return man_id

    def die_id(self):
        """
        Returns the die ID register - it should be 0x2270.
        """
        die_id = self.read_reg(REG_DIE_ID)
        return die_id

    def __del__(self):
        self.bus.close()
